<a href="https://cognitiveclass.ai/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDL0101ENSkillsNetwork20718188-2022-01-01"><img src = "https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/Logos/organization_logo/organization_logo.png" width = 400> </a>

<h1 align=center><font size = 5>Regression Models with Keras</font></h1>


## Introduction


As we discussed in the videos, despite the popularity of more powerful libraries such as PyToch and TensorFlow, they are not easy to use and have a steep learning curve. So, for people who are just starting to learn deep learning, there is no better library to use other than the Keras library.

Keras is a high-level API for building deep learning models. It has gained favor for its ease of use and syntactic simplicity facilitating fast development. As you will see in this lab and the other labs in this course, building a very complex deep learning network can be achieved with Keras with only few lines of code. You will appreciate Keras even more, once you learn how to build deep models using PyTorch and TensorFlow in the other courses.

So, in this lab, you will learn how to use the Keras library to build a regression model.


<h2>Regression Models with Keras</h2>

<h3>Objective for this Notebook<h3>    
<h5> 1. How to use the Keras library to build a regression model.</h5>
<h5> 2. Download and Clean dataset </h5>
<h5> 3. Build a Neural Network </h5>
<h5> 4. Train and Test the Network. </h5>     


## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>

1.  <a href="https://#item31">Download and Clean Dataset</a>
2.  <a href="https://#item32">Import Keras</a>
3.  <a href="https://#item33">Build a Neural Network</a>
4.  <a href="https://#item34">Train and Test the Network</a>

</font>
</div>


<a id="item31"></a>


## Download and Clean Dataset


Let's start by importing the <em>pandas</em> and the Numpy libraries.


In [1]:
# All Libraries required for this lab are listed below. The libraries pre-installed on Skills Network Labs are commented. 
# If you run this notebook on a different environment, e.g. your desktop, you may need to uncomment and install certain libraries.

#!pip install numpy==1.21.4
#!pip install pandas==1.3.4
#!pip install keras==2.1.6

In [2]:
import pandas as pd
import numpy as np

We will be playing around with the same dataset that we used in the videos.

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>


Let's download the data and read it into a <em>pandas</em> dataframe.


In [3]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa.


#### Let's check how many data points we have.


In [4]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.


Let's check the dataset for any missing values.


In [5]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [6]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.


#### Split data into predictors and target


The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.


In [7]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

<a id="item2"></a>


In [8]:
concrete_data_columns = concrete_data.columns
concrete_data_columns


Index(['Cement', 'Blast Furnace Slag', 'Fly Ash', 'Water', 'Superplasticizer',
       'Coarse Aggregate', 'Fine Aggregate', 'Age', 'Strength'],
      dtype='object')

Let's do a quick sanity check of the predictors and the target dataframes.


In [9]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [10]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.


Let's save the number of predictors to *n_cols* since we will need this number when building our network.


In [11]:
n_cols = predictors.shape[1] # number of predictors
n_cols

8

<a id="item1"></a>


<a id='item32'></a>


## Import Keras


Recall from the videos that Keras normally runs on top of a low-level library such as TensorFlow. This means that to be able to use the Keras library, you will have to install TensorFlow first and when you import the Keras library, it will be explicitly displayed what backend was used to install the Keras library. In CC Labs, we used TensorFlow as the backend to install Keras, so it should clearly print that when we import Keras.


#### Let's go ahead and import the Keras library


In [12]:
import keras
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split

As you can see, the TensorFlow backend was used to install the Keras library.


Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.


In [13]:
from keras.models import Sequential
from keras.layers import Dense

<a id='item33'></a>


## Build a Neural Network


Let's define a function that defines our regression model for us so that we can conveniently call it to create our model.


In [14]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function create a model that has one hidden layers,which contains 10 hidden units.


<a id="item4"></a>


## Train and Test the Network


**Part A**

In [15]:
# Step 1
def create_train_test(predictors,target):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size =0.3, random_state=42)
    return X_train, X_test, y_train, y_test

In [16]:
# Step 2 + 3
def eval_mod(X_train, X_test, y_train, y_test):
# build the model
    model = regression_model()
    # fit the model
    model.fit(X_train, y_train, epochs=50, verbose=0) # train on 50 epoch
    y_predict = model.predict(X_test)
    return mean_squared_error(y_predict,y_test)       # return value of mean squared error of of that model 

In [17]:
from tqdm import tqdm

n = 50
evaluate_metrics = []

for i in tqdm(range(n) ,desc = 'Loop' ):
    X_train, X_test, y_train, y_test = create_train_test(predictors, target)  # load data
    evaluate_metrics.append(eval_mod(X_train, X_test, y_train, y_test))           # return the MeanSquaredError of each model and add to list

evaluate_metrics

Loop:   0%|          | 0/50 [00:00<?, ?it/s]

10/10 [==============================] - 0s 1ms/step


Loop:   2%|▏         | 1/50 [00:02<01:46,  2.18s/it]

10/10 [==============================] - 0s 1ms/step


Loop:   4%|▍         | 2/50 [00:03<01:29,  1.86s/it]

10/10 [==============================] - 0s 1ms/step


Loop:   6%|▌         | 3/50 [00:05<01:22,  1.76s/it]

10/10 [==============================] - 0s 2ms/step


Loop:   8%|▊         | 4/50 [00:07<01:20,  1.75s/it]

10/10 [==============================] - 0s 2ms/step


Loop:  10%|█         | 5/50 [00:10<01:44,  2.33s/it]

10/10 [==============================] - 0s 2ms/step


Loop:  12%|█▏        | 6/50 [00:13<01:44,  2.39s/it]

10/10 [==============================] - 0s 3ms/step


Loop:  14%|█▍        | 7/50 [00:16<01:53,  2.65s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  16%|█▌        | 8/50 [00:17<01:38,  2.36s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  18%|█▊        | 9/50 [00:19<01:27,  2.14s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  20%|██        | 10/50 [00:21<01:19,  1.98s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  22%|██▏       | 11/50 [00:22<01:13,  1.88s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  24%|██▍       | 12/50 [00:24<01:05,  1.72s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  26%|██▌       | 13/50 [00:25<01:00,  1.63s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  28%|██▊       | 14/50 [00:27<00:58,  1.64s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  30%|███       | 15/50 [00:28<00:54,  1.56s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  32%|███▏      | 16/50 [00:30<00:53,  1.58s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  34%|███▍      | 17/50 [00:31<00:52,  1.60s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  36%|███▌      | 18/50 [00:33<00:48,  1.52s/it]

10/10 [==============================] - 0s 2ms/step


Loop:  38%|███▊      | 19/50 [00:34<00:48,  1.56s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  40%|████      | 20/50 [00:36<00:47,  1.58s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  42%|████▏     | 21/50 [00:37<00:43,  1.52s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  44%|████▍     | 22/50 [00:39<00:41,  1.47s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  46%|████▌     | 23/50 [00:40<00:41,  1.52s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  48%|████▊     | 24/50 [00:42<00:40,  1.56s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  50%|█████     | 25/50 [00:44<00:41,  1.66s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  52%|█████▏    | 26/50 [00:46<00:39,  1.65s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  54%|█████▍    | 27/50 [00:47<00:38,  1.66s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  56%|█████▌    | 28/50 [00:49<00:36,  1.66s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  58%|█████▊    | 29/50 [00:51<00:34,  1.67s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  60%|██████    | 30/50 [00:52<00:31,  1.58s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  62%|██████▏   | 31/50 [00:53<00:28,  1.52s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  64%|██████▍   | 32/50 [00:55<00:28,  1.56s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  66%|██████▌   | 33/50 [00:57<00:26,  1.58s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  68%|██████▊   | 34/50 [01:00<00:31,  1.99s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  70%|███████   | 35/50 [01:01<00:27,  1.81s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  72%|███████▏  | 36/50 [01:03<00:24,  1.76s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  74%|███████▍  | 37/50 [01:04<00:22,  1.73s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  76%|███████▌  | 38/50 [01:06<00:19,  1.62s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  78%|███████▊  | 39/50 [01:07<00:17,  1.63s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  80%|████████  | 40/50 [01:09<00:15,  1.55s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  82%|████████▏ | 41/50 [01:10<00:14,  1.57s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  84%|████████▍ | 42/50 [01:12<00:12,  1.59s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  86%|████████▌ | 43/50 [01:14<00:11,  1.60s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  88%|████████▊ | 44/50 [01:15<00:09,  1.53s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  90%|█████████ | 45/50 [01:17<00:07,  1.57s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  92%|█████████▏| 46/50 [01:18<00:06,  1.60s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  94%|█████████▍| 47/50 [01:20<00:04,  1.53s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  96%|█████████▌| 48/50 [01:21<00:03,  1.57s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  98%|█████████▊| 49/50 [01:23<00:01,  1.61s/it]

10/10 [==============================] - 0s 1ms/step


Loop: 100%|██████████| 50/50 [01:25<00:00,  1.70s/it]


[199.3090219619789,
 367.8959130174114,
 947.1599352990565,
 1306.1193903513192,
 621.6447083981839,
 916.251725793604,
 531.0419725807466,
 115.72486306144502,
 205.18668046948545,
 103.50499835751067,
 96.358814593799,
 123.64071479936742,
 129.88289660708364,
 314.7115085946102,
 666.8070927525065,
 722.2264376519812,
 141.7924349255692,
 1119.912014268595,
 105.21776273184891,
 193.14157809616793,
 1665.422912827806,
 312.2653764051414,
 122.19316142501965,
 1826.7893019199332,
 101.38125318871506,
 158.48358517041473,
 186.58204837063832,
 110.64683770662356,
 325.4359660737162,
 386.8404547247562,
 268.5841333706468,
 2060.0710300319774,
 122.72036144544789,
 457.0212760257645,
 189.79992012029163,
 364.5952075831935,
 258.3064871278513,
 130.78505110964753,
 129.38907382858736,
 267.64150518243366,
 90.19670623415786,
 337.91263997957526,
 121.93390625979987,
 221.71570089610364,
 1928.3813693676893,
 114.4881125206858,
 159.32773063128397,
 837.076002061091,
 223.76905256088028

In [18]:
print("The mean: {} and the standard deviation: {}".format(np.mean(evaluate_metrics), np.std(evaluate_metrics)))

The mean: 450.36374944714805 and the standard deviation: 505.75193475855247


**Part B** : Nomalizing data

In [19]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [20]:
n_cols_2 = predictors_norm.shape[1] # number of normalized predictors
n_cols_2

8

In [21]:
from tqdm import tqdm

n = 50
evaluate_metrics_data_norm = []

for i in tqdm(range(n) ,desc = 'Loop:' ):
    X_train, X_test, y_train, y_test = create_train_test(predictors_norm, target)  # load normalized data with preserved target
    evaluate_metrics_data_norm.append(eval_mod(X_train, X_test, y_train, y_test))  # return the MeanSquaredError of each model and add to list

evaluate_metrics_data_norm

Loop::   0%|          | 0/50 [00:00<?, ?it/s]

10/10 [==============================] - 0s 1ms/step


Loop::   2%|▏         | 1/50 [00:01<01:07,  1.37s/it]

10/10 [==============================] - 0s 1ms/step


Loop::   4%|▍         | 2/50 [00:02<01:05,  1.36s/it]

10/10 [==============================] - 0s 1ms/step


Loop::   6%|▌         | 3/50 [00:04<01:10,  1.49s/it]

10/10 [==============================] - 0s 1ms/step


Loop::   8%|▊         | 4/50 [00:05<01:06,  1.45s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  10%|█         | 5/50 [00:07<01:03,  1.42s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  12%|█▏        | 6/50 [00:08<01:05,  1.49s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  14%|█▍        | 7/50 [00:10<01:06,  1.55s/it]

10/10 [==============================] - 0s 2ms/step


Loop::  16%|█▌        | 8/50 [00:11<01:03,  1.50s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  18%|█▊        | 9/50 [00:13<01:04,  1.56s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  20%|██        | 10/50 [00:15<01:03,  1.59s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  22%|██▏       | 11/50 [00:16<01:03,  1.62s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  24%|██▍       | 12/50 [00:18<01:01,  1.62s/it]

10/10 [==============================] - 0s 2ms/step


Loop::  26%|██▌       | 13/50 [00:19<00:57,  1.55s/it]

10/10 [==============================] - 0s 2ms/step


Loop::  28%|██▊       | 14/50 [00:21<00:57,  1.59s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  30%|███       | 15/50 [00:23<00:59,  1.69s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  32%|███▏      | 16/50 [00:24<00:54,  1.59s/it]

10/10 [==============================] - 0s 2ms/step


Loop::  34%|███▍      | 17/50 [00:26<00:50,  1.54s/it]

10/10 [==============================] - 0s 2ms/step


Loop::  36%|███▌      | 18/50 [00:27<00:50,  1.59s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  38%|███▊      | 19/50 [00:29<00:50,  1.64s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  40%|████      | 20/50 [00:31<00:49,  1.64s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  42%|████▏     | 21/50 [00:32<00:47,  1.64s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  44%|████▍     | 22/50 [00:34<00:46,  1.65s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  46%|████▌     | 23/50 [00:36<00:44,  1.65s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  48%|████▊     | 24/50 [00:37<00:40,  1.56s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  50%|█████     | 25/50 [00:39<00:37,  1.50s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  52%|█████▏    | 26/50 [00:40<00:37,  1.56s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  54%|█████▍    | 27/50 [00:42<00:34,  1.51s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  56%|█████▌    | 28/50 [00:43<00:34,  1.55s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  58%|█████▊    | 29/50 [00:45<00:33,  1.58s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  60%|██████    | 30/50 [00:46<00:30,  1.53s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  62%|██████▏   | 31/50 [00:48<00:28,  1.48s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  64%|██████▍   | 32/50 [00:49<00:27,  1.53s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  66%|██████▌   | 33/50 [00:51<00:26,  1.56s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  68%|██████▊   | 34/50 [00:53<00:25,  1.58s/it]

10/10 [==============================] - 0s 2ms/step


Loop::  70%|███████   | 35/50 [00:54<00:23,  1.54s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  72%|███████▏  | 36/50 [00:56<00:21,  1.57s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  74%|███████▍  | 37/50 [00:57<00:20,  1.59s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  76%|███████▌  | 38/50 [00:59<00:18,  1.54s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  78%|███████▊  | 39/50 [01:00<00:17,  1.57s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  80%|████████  | 40/50 [01:02<00:15,  1.50s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  82%|████████▏ | 41/50 [01:03<00:13,  1.55s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  84%|████████▍ | 42/50 [01:05<00:12,  1.58s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  86%|████████▌ | 43/50 [01:06<00:10,  1.53s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  88%|████████▊ | 44/50 [01:08<00:08,  1.48s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  90%|█████████ | 45/50 [01:09<00:07,  1.53s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  92%|█████████▏| 46/50 [01:11<00:05,  1.48s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  94%|█████████▍| 47/50 [01:12<00:04,  1.54s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  96%|█████████▌| 48/50 [01:14<00:02,  1.49s/it]

10/10 [==============================] - 0s 1ms/step


Loop::  98%|█████████▊| 49/50 [01:15<00:01,  1.53s/it]

10/10 [==============================] - 0s 1ms/step


Loop:: 100%|██████████| 50/50 [01:17<00:00,  1.55s/it]


[274.839383910065,
 589.1238888631104,
 315.198606396912,
 338.93479455022583,
 370.75032633215903,
 358.57059795593136,
 331.91862578396365,
 314.26074802877946,
 284.6108267305823,
 278.25783530527707,
 714.1312308410501,
 335.62492535805603,
 495.6862280207534,
 267.38651557744413,
 243.29296953949594,
 369.58715626363306,
 309.9026350707486,
 443.0004216114633,
 306.4994382153825,
 236.96931696819428,
 236.58539386881952,
 297.4294431138128,
 332.90702011306877,
 404.8521214953019,
 380.06360221933414,
 287.75064056046847,
 234.8345319342926,
 278.2382816945093,
 243.37570429139691,
 423.7711721555203,
 206.6135726443887,
 354.97304475328036,
 322.3576258531642,
 318.1531887584238,
 386.905080690884,
 345.5501878821832,
 449.6292629102427,
 245.5931410156903,
 295.17626380755786,
 290.2002796635303,
 254.00381423778452,
 219.75875696540209,
 418.13304463304985,
 293.3236699783092,
 488.32672723643213,
 400.48916250680924,
 293.8466900258453,
 382.2015389058612,
 369.83461709802725,

In [22]:
# MSE of the normalized data

print("The normalized data has mean : {} and the standard deviation: {}".format(np.mean(evaluate_metrics_data_norm), np.std(evaluate_metrics_data_norm)))

The normalized data has mean : 339.24180897163353 and the standard deviation: 93.62555069130764


In [23]:
# MSE of the normal data (part A)
print("The mean: {} and the standard deviation: {}".format(np.mean(evaluate_metrics), np.std(evaluate_metrics)))

The mean: 450.36374944714805 and the standard deviation: 505.75193475855247


**Part C** : 100 epochs for training

In [24]:
def eval_mod_100e(X_train, X_test, y_train, y_test):
# build the model
    model = regression_model()
    # fit the model
    model.fit(X_train, y_train, epochs = 100, verbose=0) # train on 100 epochs
    y_predict = model.predict(X_test)
    return mean_squared_error(y_predict,y_test) # return the mse of model with 100epochs tranining

In [25]:
from tqdm import tqdm

n = 50
evaluate_metrics_100e = []

for i in tqdm(range(n) ,desc = 'Loop' ):
    X_train, X_test, y_train, y_test = create_train_test(predictors_norm, target)  # load normalized data
    evaluate_metrics_100e.append(eval_mod_100e(X_train, X_test, y_train, y_test))  # using model with 100 epochs training
                                 #------------#     
evaluate_metrics_100e

Loop:   0%|          | 0/50 [00:00<?, ?it/s]

10/10 [==============================] - 0s 1ms/step


Loop:   2%|▏         | 1/50 [00:02<01:53,  2.31s/it]

10/10 [==============================] - 0s 1ms/step


Loop:   4%|▍         | 2/50 [00:04<01:52,  2.35s/it]

10/10 [==============================] - 0s 1ms/step


Loop:   6%|▌         | 3/50 [00:07<02:02,  2.62s/it]

10/10 [==============================] - 0s 1ms/step


Loop:   8%|▊         | 4/50 [00:09<01:55,  2.51s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  10%|█         | 5/50 [00:12<01:59,  2.66s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  12%|█▏        | 6/50 [00:16<02:05,  2.84s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  14%|█▍        | 7/50 [00:19<02:03,  2.87s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  16%|█▌        | 8/50 [00:21<01:53,  2.71s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  18%|█▊        | 9/50 [00:24<01:53,  2.77s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  20%|██        | 10/50 [00:27<01:52,  2.82s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  22%|██▏       | 11/50 [00:30<01:51,  2.86s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  24%|██▍       | 12/50 [00:32<01:42,  2.71s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  26%|██▌       | 13/50 [00:34<01:35,  2.59s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  28%|██▊       | 14/50 [00:37<01:30,  2.52s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  30%|███       | 15/50 [00:40<01:32,  2.64s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  32%|███▏      | 16/50 [00:43<01:32,  2.73s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  34%|███▍      | 17/50 [00:45<01:26,  2.62s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  36%|███▌      | 18/50 [00:47<01:21,  2.54s/it]

10/10 [==============================] - 0s 2ms/step


Loop:  38%|███▊      | 19/50 [00:50<01:22,  2.66s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  40%|████      | 20/50 [00:53<01:21,  2.73s/it]

10/10 [==============================] - 0s 2ms/step


Loop:  42%|████▏     | 21/50 [00:55<01:16,  2.62s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  44%|████▍     | 22/50 [00:58<01:10,  2.52s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  46%|████▌     | 23/50 [01:00<01:06,  2.48s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  48%|████▊     | 24/50 [01:03<01:08,  2.63s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  50%|█████     | 25/50 [01:06<01:08,  2.72s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  52%|█████▏    | 26/50 [01:09<01:04,  2.71s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  54%|█████▍    | 27/50 [01:11<01:00,  2.62s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  56%|█████▌    | 28/50 [01:14<00:59,  2.71s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  58%|█████▊    | 29/50 [01:17<00:55,  2.62s/it]

10/10 [==============================] - 0s 2ms/step


Loop:  60%|██████    | 30/50 [01:20<00:54,  2.74s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  62%|██████▏   | 31/50 [01:23<00:53,  2.83s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  64%|██████▍   | 32/50 [01:25<00:48,  2.69s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  66%|██████▌   | 33/50 [01:27<00:43,  2.59s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  68%|██████▊   | 34/50 [01:30<00:42,  2.69s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  70%|███████   | 35/50 [01:33<00:41,  2.76s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  72%|███████▏  | 36/50 [01:36<00:39,  2.80s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  74%|███████▍  | 37/50 [01:38<00:34,  2.69s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  76%|███████▌  | 38/50 [01:41<00:33,  2.76s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  78%|███████▊  | 39/50 [01:44<00:30,  2.81s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  80%|████████  | 40/50 [01:47<00:28,  2.84s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  82%|████████▏ | 41/50 [01:50<00:24,  2.71s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  84%|████████▍ | 42/50 [01:52<00:20,  2.60s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  86%|████████▌ | 43/50 [01:54<00:17,  2.52s/it]

10/10 [==============================] - 0s 2ms/step


Loop:  88%|████████▊ | 44/50 [01:57<00:15,  2.66s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  90%|█████████ | 45/50 [02:00<00:13,  2.74s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  92%|█████████▏| 46/50 [02:03<00:11,  2.80s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  94%|█████████▍| 47/50 [02:06<00:08,  2.76s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  96%|█████████▌| 48/50 [02:09<00:05,  2.82s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  98%|█████████▊| 49/50 [02:11<00:02,  2.68s/it]

10/10 [==============================] - 0s 1ms/step


Loop: 100%|██████████| 50/50 [02:14<00:00,  2.69s/it]


[155.48728411140877,
 144.72518488514183,
 175.4696943520169,
 134.68913783702962,
 149.32136330555682,
 156.43694712062867,
 163.48612945386296,
 173.10997280381676,
 183.16721370786732,
 178.60961432468514,
 143.89991065743226,
 143.24464062970068,
 162.82025017675048,
 158.6433776889475,
 149.7424651257672,
 181.14452124296884,
 163.3760167828557,
 150.09955556150945,
 148.7820953903392,
 150.0347658668863,
 157.10747374633522,
 149.96701238373768,
 145.22988654692264,
 148.17881680751233,
 144.98082933199768,
 153.80796643001926,
 140.17741515720763,
 154.49234375632366,
 158.83604234713795,
 139.5284324214354,
 161.65088507010535,
 140.53175266963655,
 152.80419323732875,
 177.7797111028024,
 166.74010360606943,
 151.16422367386255,
 162.85528781348359,
 142.20018867728328,
 154.4266515041291,
 171.56330109142922,
 141.76579541906023,
 142.18950308924394,
 141.74083334349714,
 152.78816353426438,
 159.70402438542206,
 145.87869404947048,
 165.97617360335943,
 163.09870955067987,
 

In [26]:
# MSE of model with 100 epochs training

print("The evaluate_metrics_100e list has mean : {} and the standard deviation: {}".format(np.mean(evaluate_metrics_100e), np.std(evaluate_metrics_100e)))

The evaluate_metrics_100e list has mean : 154.89640013768334 and the standard deviation: 11.932508439653535


In [27]:
# Compare with MSE of the normalized data (part B) 

print("The normalized data has mean : {} and the standard deviation: {}".format(np.mean(evaluate_metrics_data_norm), np.std(evaluate_metrics_data_norm)))

The normalized data has mean : 339.24180897163353 and the standard deviation: 93.62555069130764


Part D: using different regession model

In [28]:
# define regression model
def regression_model_2():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [29]:
def eval_mod_diff_model(X_train, X_test, y_train, y_test):
# build the model
    model = regression_model_2() # using model with different architecture
    # fit the model
    model.fit(X_train, y_train, epochs=50, verbose=0) # train on 50 epoch
    y_predict = model.predict(X_test)
    return mean_squared_error(y_predict,y_test) 

In [30]:
from tqdm import tqdm

n = 50
evaluate_metrics_model_2 = []

for i in tqdm(range(n) ,desc = 'Loop' ):
    X_train, X_test, y_train, y_test = create_train_test(predictors_norm, target)  # load normalized data
    evaluate_metrics_model_2.append(eval_mod_diff_model(X_train, X_test, y_train, y_test))  
                                    # using model with different architecture

evaluate_metrics_model_2

Loop:   0%|          | 0/50 [00:00<?, ?it/s]

10/10 [==============================] - 0s 1ms/step


Loop:   2%|▏         | 1/50 [00:01<01:25,  1.74s/it]

10/10 [==============================] - 0s 1ms/step


Loop:   4%|▍         | 2/50 [00:03<01:20,  1.68s/it]

10/10 [==============================] - 0s 1ms/step


Loop:   6%|▌         | 3/50 [00:05<01:18,  1.68s/it]

10/10 [==============================] - 0s 1ms/step


Loop:   8%|▊         | 4/50 [00:06<01:18,  1.70s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  10%|█         | 5/50 [00:08<01:16,  1.69s/it]

10/10 [==============================] - 0s 2ms/step


Loop:  12%|█▏        | 6/50 [00:10<01:13,  1.68s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  14%|█▍        | 7/50 [00:11<01:12,  1.68s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  16%|█▌        | 8/50 [00:13<01:10,  1.68s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  18%|█▊        | 9/50 [00:15<01:08,  1.67s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  20%|██        | 10/50 [00:16<01:06,  1.66s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  22%|██▏       | 11/50 [00:18<01:05,  1.68s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  24%|██▍       | 12/50 [00:20<01:03,  1.67s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  26%|██▌       | 13/50 [00:22<01:04,  1.74s/it]

10/10 [==============================] - 0s 2ms/step


Loop:  28%|██▊       | 14/50 [00:23<01:01,  1.71s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  30%|███       | 15/50 [00:25<00:59,  1.71s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  32%|███▏      | 16/50 [00:27<00:58,  1.72s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  34%|███▍      | 17/50 [00:28<00:55,  1.69s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  36%|███▌      | 18/50 [00:30<00:54,  1.70s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  38%|███▊      | 19/50 [00:32<00:52,  1.69s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  40%|████      | 20/50 [00:33<00:50,  1.68s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  42%|████▏     | 21/50 [00:35<00:48,  1.68s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  44%|████▍     | 22/50 [00:37<00:47,  1.68s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  46%|████▌     | 23/50 [00:38<00:45,  1.70s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  48%|████▊     | 24/50 [00:40<00:44,  1.70s/it]

10/10 [==============================] - 0s 2ms/step


Loop:  50%|█████     | 25/50 [00:42<00:42,  1.70s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  52%|█████▏    | 26/50 [00:44<00:41,  1.73s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  54%|█████▍    | 27/50 [00:45<00:39,  1.70s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  56%|█████▌    | 28/50 [00:47<00:39,  1.79s/it]

10/10 [==============================] - 0s 2ms/step


Loop:  58%|█████▊    | 29/50 [00:49<00:37,  1.78s/it]

10/10 [==============================] - 0s 2ms/step


Loop:  60%|██████    | 30/50 [00:51<00:35,  1.77s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  62%|██████▏   | 31/50 [00:52<00:32,  1.73s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  64%|██████▍   | 32/50 [00:54<00:31,  1.73s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  66%|██████▌   | 33/50 [00:56<00:29,  1.74s/it]

10/10 [==============================] - 0s 2ms/step


Loop:  68%|██████▊   | 34/50 [00:58<00:27,  1.73s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  70%|███████   | 35/50 [00:59<00:25,  1.72s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  72%|███████▏  | 36/50 [01:01<00:24,  1.73s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  74%|███████▍  | 37/50 [01:03<00:22,  1.73s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  76%|███████▌  | 38/50 [01:05<00:20,  1.73s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  78%|███████▊  | 39/50 [01:06<00:18,  1.70s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  80%|████████  | 40/50 [01:08<00:17,  1.71s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  82%|████████▏ | 41/50 [01:10<00:15,  1.71s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  84%|████████▍ | 42/50 [01:11<00:13,  1.70s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  86%|████████▌ | 43/50 [01:13<00:12,  1.77s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  88%|████████▊ | 44/50 [01:15<00:10,  1.73s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  90%|█████████ | 45/50 [01:17<00:09,  1.86s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  92%|█████████▏| 46/50 [01:19<00:07,  1.79s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  94%|█████████▍| 47/50 [01:20<00:05,  1.78s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  96%|█████████▌| 48/50 [01:22<00:03,  1.76s/it]

10/10 [==============================] - 0s 1ms/step


Loop:  98%|█████████▊| 49/50 [01:24<00:01,  1.75s/it]

10/10 [==============================] - 0s 1ms/step


Loop: 100%|██████████| 50/50 [01:26<00:00,  1.72s/it]


[113.9739036121017,
 131.53668734843956,
 86.18479850364788,
 134.23928540115043,
 126.70996145014773,
 139.9233690261514,
 133.5530724640715,
 124.01947413667276,
 128.58201506496405,
 135.23427640933792,
 110.87617781957543,
 127.26549362020786,
 134.37625322863718,
 132.2344570321703,
 132.58669896952634,
 126.7153669166385,
 126.59465068299875,
 124.22638836493456,
 133.58305752815286,
 125.24777975495061,
 136.08741873048663,
 106.65305013695874,
 132.8131231835055,
 117.71807137533814,
 97.48698809971211,
 133.98565015052392,
 128.7310134077815,
 116.72125944061564,
 139.60304550668405,
 111.22360087459523,
 133.18610513216447,
 131.6749257252029,
 128.7191078377437,
 105.99505214405569,
 133.44605248053875,
 104.90702173382893,
 134.69261157826813,
 128.79245297917157,
 129.9536248039242,
 125.02424389766198,
 82.66325766583914,
 130.21196638393485,
 126.71851807361436,
 123.4244618972667,
 130.48519964417994,
 134.28087920447794,
 109.0939774865502,
 129.5192374541022,
 119.242

In [31]:
print("The evaluate_metrics of different model has mean : {} and the standard deviation: {}".format(np.mean(evaluate_metrics_model_2), np.std(evaluate_metrics_model_2)))

The evaluate_metrics of different model has mean : 124.10373019019106 and the standard deviation: 12.621974753137275


In [32]:
# Compare with MSE of the normalized data (part B) 

print("The normalized data has mean : {} and the standard deviation: {}".format(np.mean(evaluate_metrics_data_norm), np.std(evaluate_metrics_data_norm)))

The normalized data has mean : 339.24180897163353 and the standard deviation: 93.62555069130764


<strong>You can refer to this [link](https://keras.io/models/sequential/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDL0101ENSkillsNetwork20718188-2022-01-01) to learn about other functions that you can use for prediction or evaluation.</strong>


Feel free to vary the following and note what impact each change has on the model's performance:

1.  Increase or decreate number of neurons in hidden layers
2.  Add more hidden layers
3.  Increase number of epochs


### Thank you for completing this lab!

This notebook was created by [Alex Aklson](https://www.linkedin.com/in/aklson/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDL0101ENSkillsNetwork20718188-2022-01-01). I hope you found this lab interesting and educational. Feel free to contact me if you have any questions!


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description                                          |
| ----------------- | ------- | ---------- | ----------------------------------------------------------- |
| 2020-09-21        | 2.0     | Srishti    | Migrated Lab to Markdown and added to course repo in GitLab |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>


This notebook is part of a course on **Coursera** called *Introduction to Deep Learning & Neural Networks with Keras*. If you accessed this notebook outside the course, you can take this course online by clicking [here](https://cocl.us/DL0101EN_Coursera_Week3\_LAB1).


<hr>

Copyright © 2019 [IBM Developer Skills Network](https://cognitiveclass.ai/?utm_medium=dswb&utm_source=bducopyrightlink&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDL0101ENSkillsNetwork20718188-2022-01-01&utm_campaign=bdu). This notebook and its source code are released under the terms of the [MIT License](https://bigdatauniversity.com/mit-license/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDL0101ENSkillsNetwork20718188-2022-01-01).
